metadata-ingestion/src/datahub/ingestion/source/dbt/dbt_common.py

```diff
@@ -707,6 +850,182 @@ def filter_nodes(self, all_nodes: List[DBTNode]) -> List[DBTNode]:
+
+    def _infer_schemas_and_update_cll(self, all_nodes_map: Dict[str, DBTNode]) -> None:
+        if not self.config.infer_dbt_schemas:
+            if self.config.include_column_lineage:
+                raise ConfigurationError(
+                    "`infer_dbt_schemas` must be enabled to use `include_column_lineage`"
+                )
+            return
+
+        graph = self.ctx.graph
+
+        schema_resolver = SchemaResolver(
+            platform=self.config.target_platform,
+            platform_instance=self.config.target_platform_instance,
+            env=self.config.env,
+        )
+
+        target_platform_urn_to_dbt_name: Dict[str, str] = {}
+
+        # Iterate over the dbt nodes in topological order.
+        # This ensures that we process upstream nodes before downstream nodes.
+        for dbt_name in topological_sort(
+            list(all_nodes_map.keys()),
+            edges=list(
+                itertools.chain.from_iterable(
+                    [(upstream, node.dbt_name) for upstream in node.upstream_nodes]
+                    for node in all_nodes_map.values()
+                )
+            ),
+        ):
+            node = all_nodes_map[dbt_name]
+
+            target_node_urn = None
+            should_fetch_target_node_schema = False
+            if node.exists_in_target_platform:
+                target_node_urn = node.get_urn(
+                    self.config.target_platform,
+                    self.config.env,
+                    self.config.target_platform_instance,
+                )
+                should_fetch_target_node_schema = True
+            elif node.is_ephemeral_model():
+                # For ephemeral nodes, we "pretend" that they exist in the target platform
+                # for schema resolution purposes.
+                target_node_urn = mce_builder.make_dataset_urn_with_platform_instance(
+                    platform=self.config.target_platform,
+                    name=node.get_fake_ephemeral_table_name(),
+                    platform_instance=self.config.target_platform_instance,
+                    env=self.config.env,
+                )
+            if target_node_urn:
+                target_platform_urn_to_dbt_name[target_node_urn] = node.dbt_name
+
+            # Our schema resolver preference is:
+            # 1. graph
+            # 2. dbt catalog
+            # 3. inferred
+            # Exception: if convert_column_urns_to_lowercase is enabled, swap 1 and 2.
+            # Cases 1 and 2 are handled here, and case 3 is handled after schema inference has occurred.
+            schema_fields: Optional[List[SchemaField]] = None
+
+            # Fetch the schema from the graph.
+            if (
+                not schema_fields
+                and target_node_urn
+                and should_fetch_target_node_schema
+                and graph
+            ):
+                schema_metadata = graph.get_aspect(target_node_urn, SchemaMetadata)
```

Comment:
This isn't cached right? Is it possible we end up querying this multiple times for the same urn? Could we do a bulk fetch instead?

Reply:
it won't ever query for the same urn multiple times